### ISSUE TITLE
Python SDK: Not all the decorators are working

### ISSUE THREAD

**Issue by jim-hm** [dagger/dagger#8698]
### What is the issue?

My function failed when I tried to add decorators on it. I can find workaround but not sure if it should add to docs.

 

### Dagger version

dagger v0.13.3 (registry.dagger.io/engine:v0.13.3) darwin/arm64

### Steps to reproduce

get following error when run `dagger call load-context` (module name is test)

`
def timer(func):
    """Print the runtime of the decorated function"""
    # @functools.wraps(func)
    async def wrapper_timer(*args, **kwargs):
        start_time = time.perf_counter()
        value = await func(*args, **kwargs)
        end_time = time.perf_counter()
        run_time = end_time - start_time
        print(f"Finished {func.__name__}() in {run_time:.4f} secs")
        return value
    return wrapper_timer

@object_type
class Test:

    @function
    @timer
    async def load_context(self) -> str:
        """
        Returns the contents of the pyproject.toml file in the current module's source directory
        """
        file =  dag.current_module().source().file('pyproject.toml')
        rs = await file.contents()
        return rs
`

### Log output

✔ connect 0.3s
✘ initialize 4.0s
! get module name: input: failed to get schema: failed to get schema for module "test": failed to create function "loadContext": failed to find mod type for function "loadContext" arg "args" type
  ✔ resolving module ref 0.1s
  ✔ installing module 3.9s
  ✘ analyzing module 0.0s
  ! get module name: input: failed to get schema: failed to get schema for module "test": failed to create function "loadContext": failed to find mod type for function "loadContext" arg "args" type

**Comment by jim-hm** [comment]
if uncomment # @functools.wraps(func), the decorator is working. it seems dagger decorator depends on the meta data of original function?

**Comment by helderco** [comment]
At first glance, you should have `@function` closer to your function, like this:
```python
@timer
@function
async def load_context(self) -> str:
    ...
```
If it still doesn't work, I'll take a look when I can.

**Comment by jim-hm** [comment]
Thanks @helderco , 
move @function close to function did not help. Here's error, 
Error: input: module.withSource.initialize resolve: failed to initialize module: failed to call module "test" to get functions: call constructor: process "/runtime" did not complete successfully: exit code: 1

Stderr:
╭─ Error ──────────────────────────────────────────────────────────────────────╮
│ Function “Test.load_context” seems to be decorated in a class that is not    │
│ itself decorated with @object_type                                           │
╰──────────────────────────────────────────────────────────────────────────────╯

I also add typing info, it seems nothing changed. I still need annotate func by @functools.wraps(func) to get it working.

**Comment by helderco** [comment]
> ```
> ╭─ Error ──────────────────────────────────────────────────────────────────────╮
> │ Function “Test.load_context” seems to be decorated in a class that is not    │
> │ itself decorated with @object_type                                           │
> ╰──────────────────────────────────────────────────────────────────────────────╯
> ```

Right, that's because your decorator is replacing the descriptor that `@function` returns, so it's not updating its "origin" properly to the class, which fails validation because it looks like the function is floating. I'm actually working on simplifying all of this which should make your use case work or at least easier to customize in:
- https://github.com/dagger/dagger/pull/8709

In the meantime, there's ways to get around it. The `func` that you get in your decorator should be an instance of the `dagger.mod._resolver:Function` descriptor. You can extend that class to hook your timing logic in, but you need to return it back from your decorator so that its `__set_name__` method is able to set things up correctly, especially this:

https://github.com/dagger/dagger/blob/05af295351e2d9154d7b72c06443f8152a992680/sdk/python/src/dagger/mod/_resolver.py#L437

> I also add typing info, it seems nothing changed. I still need annotate func by @functools.wraps(func) to get it working. Here's my typings `async def wrapper_timer(*args: Tuple[Any, ...], **kwargs: Dict[str, Any]) -> Any:`

To be clear, the thing about types is that it matters to `@function` which types the decorated function uses, since it inspects the function's signature in order to report the right things to the Dagger API. Changing the typing on the wrapper that's outside of `@function` won't have any effect, and if it's below it it won't work becuase those types are general and dynamic, and don't match the underlying `load_context` function's signature at all.

**Comment by jim-hm** [comment]
Thank you for explanation. If my understanding is correct the recommendation is that I should always move `@function` close to original function.
>> In the meantime, there's ways to get around it. The func that you get in your decorator should be an instance of the dagger.mod._resolver:Function descriptor. You can extend that class to hook your timing logic in, but you need to return it back from your decorator so that its __set_name__ method is able to set things up correctly, especially this:
It seems `dagger.mod._resolver:Function` is not open for other packages. How I can cast `func` to a instance of it?
